src/main/cpp/include/query_operations/pl_remapper_template_definition.h

```diff
@@ -0,0 +1,247 @@
+    m_merged_to_input_allele_idx[i] =
+        m_alleles_combiner->template get_input_allele_idx<do_remap, is_REF_block, contains_NON_REF_allele>(
+            row_query_idx, i);
+  auto num_gts = KnownFieldInfo::get_number_of_genotypes(num_merged_alleles - 1u, ploidy);  //-1 for #ALT alleles
+  // First element - all elements of ploidy == REF
+  auto input_gt_idx = 0;
+  if (static_cast<unsigned>(input_gt_idx) >= length)
+    no_overflow = no_overflow && op.template write_missing<element_ty, true>(m_PL_query_idx, row_query_idx);
+  else
+    no_overflow = no_overflow && op.template write_element<element_ty, true>(m_PL_query_idx, row_query_idx,
+                                                                             PL_field_ptr[input_gt_idx]);
+  // FIXME: genotype count limits
+  switch (ploidy_enum) {
+    case PloidyEnum::HAPLOID:  // haploid, number of genotypes == number of alleles
+    {
+      for (auto i = 1u; i < num_gts; ++i) {
+        auto input_gt_idx = m_merged_to_input_allele_idx[i];
+        if (input_gt_idx < 0 || static_cast<unsigned>(input_gt_idx) >= length)
+          no_overflow = no_overflow && op.template write_missing<element_ty, false>(m_PL_query_idx, row_query_idx);
+        else
+          no_overflow = no_overflow && op.template write_element<element_ty, false>(m_PL_query_idx, row_query_idx,
+                                                                                    PL_field_ptr[input_gt_idx]);
+      }
+      break;
+    }
+    case PloidyEnum::DIPLOID: {
+      for (auto allele_j = 0u; allele_j < num_merged_alleles; ++allele_j) {
+        auto input_allele_j = m_merged_to_input_allele_idx[allele_j];
+        if (input_allele_j < 0) {  // merged allele has no counterpart in input (including NON_REF)
+          for (auto allele_k = (allele_j == 0u) ? 1u : allele_j; allele_k < num_merged_alleles; ++allele_k) {
+            no_overflow = no_overflow && op.template write_missing<element_ty, false>(m_PL_query_idx, row_query_idx);
+          }
+          continue;
+        }
+        // First element of PL vector has been dealt outside the switch statement
+        for (auto allele_k = (allele_j == 0u) ? 1u : allele_j; allele_k < num_merged_alleles; ++allele_k) {
+          auto input_allele_k = m_merged_to_input_allele_idx[allele_k];
+          auto input_gt_idx = bcf_alleles2gt(input_allele_j, input_allele_k);
+          if (input_gt_idx < 0 || static_cast<unsigned>(input_gt_idx) >= length)
+            no_overflow = no_overflow && op.template write_missing<element_ty, false>(m_PL_query_idx, row_query_idx);
+          else
+            no_overflow = no_overflow && op.template write_element<element_ty, false>(m_PL_query_idx, row_query_idx,
+                                                                                      PL_field_ptr[input_gt_idx]);
+        }
+      }
+      break;
+    }
+    case PloidyEnum::GENERAL: {
+#define SET_PLOIDY_INDEX_IN_STACK_ELEMENT(X, Y) ((X).first = (Y))
```

Comment:
It's a bit confusing to see pragma definitions inside a case statement, can they be pulled outside the function?